aten/src/ATen/native/vulkan/api/Resource.h

```diff
@@ -354,21 +394,29 @@ inline Resource::Fence::operator bool() const {
   return pool;
 }
 
-inline VkFence Resource::Fence::handle(const bool add_to_waitlist) const {
-  if (!pool) {
-    return VK_NULL_HANDLE;
-  }
-
-  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(
-      id < pool->fence_.pool.size(),
-      "Invalid Vulkan fence!");
-
-  const VkFence fence = pool->fence_.pool[id].get();
-  if (add_to_waitlist) {
-    pool->fence_.waitlist.push_back(fence);
+template<typename Block>
+inline Resource::Buffer Resource::Pool::uniform(const Block& block) {
+  Buffer uniform = this->buffer({
+      sizeof(Block),
+      {
+        VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT,
+        {
+          VMA_MEMORY_USAGE_CPU_TO_GPU,
+          0u,
+          0u,
+        },
+      },
+    });
+
+  {
+    Memory::Handle<Block*> memory = uniform.memory.template map<
+        Block,
+        Memory::Access::Write>();
+
+    *memory.get() = block;
```

Comment:
Am I right that semantic of type Block here is  something that has sizeof and and can be assigned to memory pointer. Will it copy here all the block data to allocated memory?

Reply:
Yes, almost.  Invocation of sizeof on any data type or variable in C++ is valid.  As for the second part of your statement, if Block has an assignment operator defined, whether implicitly by the compiler such as the cases where Block is a simple POD as is all of our use cases, or explicitly defined by the user if Block has more complex semantics, then this code will compile.  The reason behind this implementation is two fold:  The primary motivation is that memcpy is a function call unless treated specifically by the compiler as an intrinsic considering how common place it is, and calling a complex function to copy a few bytes is overkill.  The second, in this particular case less important, reason is that memcpy is only technically a valid operation on PODs, otherwise the assignment operator should be called for proper treatment of non primitive types.

In the future we should move to push constants where possible.  That will inline parameter passing (for small blocks) inside the command buffer and should be much faster compared to all the overhead of allocation a uniform buffer (even from a pool as do), creation of vkBuffer, mapping and unmapping of that memory, all for a few bytes of data.